internal/sequencers/espresso/espresso_client.go

```diff
@@ -0,0 +1,268 @@
+	Account     string          `json:"account"`
+	Domain      EIP712Domain    `json:"domain"`
+	Types       Types           `json:"types"`
+	PrimaryType string          `json:"primaryType"`
+	Message     EspressoMessage `json:"message"`
+}
+
+const (
+	HARDHAT         = 31337
+	PURPOSE_INDEX   = 44
+	COIN_TYPE_INDEX = 60
+)
+
+func (e *EspressoClient) SendInput(payload string, namespace int) {
+	mnemonic := "test test test test test test test test test test test junk"
+	client, err := ethclient.Dial("http://localhost:8545")
+	if err != nil {
+		log.Fatalf("Error connection to the node: %v", err)
+	}
+
+	// Generate private key from mnemonic
+	privateKey, err := getPrivateKeyFromMnemonic(mnemonic)
+	if err != nil {
+		log.Fatalf("Erro ao derivar chave privada: %v", err)
+	}
+
+	// Sign and Send Espresso Input
+	response, err := addEspressoInput(e, client, privateKey, namespace, payload)
+	if err != nil {
+		log.Fatalf("Error sending to Espresso: %v", err)
+	}
+	fmt.Println("Transaction received:", response)
+}
+
+func addEspressoInput(e *EspressoClient, client *ethclient.Client, privateKey *ecdsa.PrivateKey, namespace int, payload string) (string, error) {
+	// Get nonce
+	nonce, err := fetchNonce(client, privateKey)
+	if err != nil {
+		return "", fmt.Errorf("erro ao buscar nonce: %v", err)
+	}
+
+	fromAddress := crypto.PubkeyToAddress(privateKey.PublicKey)
+
+	espressoMessage := EspressoMessage{
+		Nonce:   nonce,
+		Payload: "0x" + payload,
+	}
+
+	// Build Message
+	typedData := EspressoData{
+		Account: fromAddress.Hex(),
+		Message: espressoMessage,
+		Domain: EIP712Domain{
+			Name:              "EspressoM",
+			Version:           "1",
+			ChainId:           HARDHAT,
+			VerifyingContract: "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
+		},
+		PrimaryType: "EspressoMessage",
+		Types: Types{
+			EIP712Domain: []TypeDetail{
+				{Name: "name", Type: "string"},
+				{Name: "version", Type: "string"},
+				{Name: "chainId", Type: "uint32"},
+				{Name: "verifyingContract", Type: "address"},
+			},
+			EspressoMessage: []TypeDetail{
+				{Name: "nonce", Type: "uint64"},
+				{Name: "payload", Type: "string"},
+			},
+		},
+	}
+
+	// Sign typed data
+	signature, err := signTypedData(privateKey, typedData)
+
+	if err != nil {
+		return "", fmt.Errorf("Error getting signature: %v", err)
+	}
+
+	signedMessage, err := createSignedMessage(signature, typedData)
+
+	if err != nil {
+		return "", fmt.Errorf("Error signing message: %v", err)
+	}
+
+	// Send signed message to Espresso
+	response, err := submitToEspresso(e, namespace, signedMessage)
+	if err != nil {
+		return "", fmt.Errorf("Error sending input to Espresso: %v", err)
+	}
+
+	fmt.Println("Input sent to Espresso successfully!")
+	return response, nil
+}
+
+func fetchNonce(client *ethclient.Client, privateKey *ecdsa.PrivateKey) (uint64, error) {
+	fromAddress := crypto.PubkeyToAddress(privateKey.PublicKey)
+	nonce, err := client.PendingNonceAt(context.Background(), fromAddress)
```

Comment:
This looks strange honestly. How do you use `ethclient` to fetch nonce and use it as Espresso nonce? They are 2 different things in my opinion.
In the browser front end, we query graphql for the nonce like [this](https://github.com/ZzzzHui/frontend-web-cartesi/blob/24fa6d1b3f43f29ea716c27a5c2b7c04b0d2bb94/src/Input.tsx#L87). Feel free to let me know what you think :)

Reply:
Hey @ZzzzHui  you are right. Updated the code to match your example.